Implement support for the free-threaded build of CPython 3.13

lysnikolaou opened the pull request:
<!-- Thank you for your contribution! -->

## What do these changes do?

- Add a matrix config for free-threaded Python to run tests with it.
- Change configuration to correctly build wheels under the free-threaded build.


<!-- Please give a short brief about these changes. -->

## Are there changes in behavior for the user?

No.

<!-- Outline any notable behaviour for the end users. -->

## Related issue number

<!-- Are there any issues opened that will be resolved by merging this change? -->
<!-- Remember to prefix with 'Fixes' if it should close the issue (e.g. 'Fixes #123'). -->

## Checklist

- [x] I think the code is well written
- [x] Unit tests for the changes exist
- [ ] Documentation reflects the changes

bdraco commented on .github/workflows/ci-cd.yml:
```diff
@@ -231,7 +232,7 @@ jobs:
     - name: Setup Python ${{ matrix.pyver }}
       id: python-install
-      uses: actions/setup-python@v5
+      uses: quansight-labs/setup-python@v5
```
Can you provide some more color about this change?

Dreamsorcerer replied:
Looks like this: https://github.com/actions/setup-python/pull/973#issuecomment-2495900996

Can we again add a TODO comment to revert this once resolved: https://github.com/actions/setup-python/issues/771.

Dreamsorcerer commented on .github/workflows/ci-cd.yml:
```diff
@@ -241,6 +242,14 @@ jobs:
       uses: py-actions/py-dependency-install@v4
       with:
         path: requirements/test.txt
+    - name: Install Cython nightly on free-threading
```
I assume this can also be removed once Cython 3.1 is out? Can we get a TODO comment here as well.

Dreamsorcerer replied:
Actually, will 3.1.0a1 work? Would probably simplify some of this, as that's already on PyPI.

webknjaz replied:
Wasn't this the one that broke yarl during the sprints?

bdraco replied:
Yes

lysnikolaou replied:
3.1.0a1 would work sufficiently well, though there's many Cython improvements that are being currently worked on and will appear in the nightlies much sooner.

Dreamsorcerer replied:
I think to ease future maintenance, it'd be best to use a released version so we don't need to update this again later and can keep it simple. If there's major improvements we want, we can probably just ask them for another alpha release.

lysnikolaou replied:
Sure! Keep in mind that Cython doesn't really do many releases though.

Dreamsorcerer commented on pyproject.toml:
```diff
@@ -78,6 +79,11 @@ test-command = "pytest -v --no-cov {project}/tests"
 # don't build PyPy wheels, install from source instead
 skip = "pp*"
 
+[[tool.cibuildwheel.overrides]]
+select = "cp313t-*"
+build-frontend = "build; args: --no-isolation"
+before-build = "bash {package}/scripts/cibw_before_build.sh"
+
```
Also another comment here, I'm unclear if only the before-build line or other lines should be removed when Cython 3.1 is out.

webknjaz replied:
Isn't this a list? Perhaps, the script could be stuffed right inside this config?

lysnikolaou replied:
> Isn't this a list? Perhaps, the script could be stuffed right inside this config?

Not sure whether that would work cross-platform, but I can try it if you want.

Dreamsorcerer commented on MANIFEST.in:
```diff
@@ -12,6 +12,7 @@ graft docs
 graft CHANGES
 graft requirements
 graft tests
+graft scripts
```
This could also be reverted, so could do with a comment as well.

webknjaz commented on scripts/cibw_before_build.sh:
```diff
@@ -0,0 +1,7 @@
+FREE_THREADED_BUILD="$(python -c"import sysconfig; print(bool(sysconfig.get_config_var('Py_GIL_DISABLED')))")"
+if [[ $FREE_THREADED_BUILD == "True" ]]; then
+    python -m pip install -U pip
+    python -m pip install -i https://pypi.anaconda.org/scientific-python-nightly-wheels/simple cython
```
This should be possible to implement less intrusively though `PIP_CONSTRAINT` and `Cython @ url` syntax in the constraint file..

lysnikolaou replied:
That would mean that we install a nightly Cython for non-free-threaded Pythons as well though, which to me seems suboptimal. What do you think?

lysnikolaou commented on src/propcache/_helpers_c.pyx:
```diff
@@ -1,4 +1,4 @@
-# cython: language_level=3
+# cython: language_level=3, freethreading_compatible=True
```
I tried that but it unfortunately cannot, because there's not an option to add version-specific option to pyproject.toml. For this, the build backend would have to be changed, which seems like a bigger hassle.

webknjaz replied:
Isn't this one version-agnostic either?

lysnikolaou commented:
Hey folks, thanks for all the review and sorry for the extended delay here, I was out on PTO. I've addressed most of the comments around adding some TODOs and replied to the rest.

lysnikolaou commented:
Gentle ping on this. Is there anything I can do to move it forward?

bdraco commented on packaging/pep517_backend/_backend.py:
```diff
@@ -372,7 +372,8 @@ def get_requires_for_build_wheel(
 
     c_ext_build_deps = [] if is_pure_python_build else [
-        'Cython ~= 3.0.0; python_version >= "3.12"',
+        'Cython == 3.1.0a1; python_version >= "3.13"',
```
Looks like the performance regression is coming from 3.1.0a1.

We should only use 3.1.0a1 if its the free-threading 3.13 so this doesn't affect the current 3.13 production builds

lysnikolaou replied:
> Looks like the performance regression is coming from 3.1.0a1.
> 
> We should only use 3.1.0a1 if its the free-threading 3.13 so this doesn't affect the current 3.13 production builds

That's not very easy to do if we continue using a constraint file for Cython since there's no marker for the free-threaded build. It seems though that we could remove the contraints from a couple of places:

- When building sdists and pure python wheels: Cython is probably not needed there.
- For the configuration environment in cibuildwheel: Can be removed, since the Cython versions that ends up being installed is the one specified in the build backend. 

I'm not sure how we would do that for the test requirements though, since Cython is needed there as well. We would have to use something more complicated to be able to install the correct version.

bdraco replied:
@webknjaz would you kindly offer some guidance here?

webknjaz replied:
@bdraco oh, sorry. I missed this discussion. In general, it'd be good to have loose versioning here, but pinned in the constraint file. This mostly influences what's in the isolated build env, but some people (downstreams?) can disable isolation. And also when the callers set their own constraint files, this may contribute to conflicts during dependency resolution. So I usually treat the build deps as normal runtime packaging metadata deps with the option to pin them through the constraints mechanism in pip.

bdraco reviewed (changes requested):
@lysnikolaou sorry for the delay

I think this is good to go once https://github.com/aio-libs/propcache/pull/84/files#r1925771536 is addressed

webknjaz commented on packaging/pep517_backend/_backend.py:
```diff
@@ -371,10 +372,13 @@ def get_requires_for_build_wheel(
+        c_ext_build_deps = []
+    elif sysconfig.get_config_var("Py_GIL_DISABLED"):
+        # TODO: Remove when there's a Cython final with free-threading support
+        c_ext_build_deps = ['Cython == 3.1.0a1']
```
Why not compat?
```suggestion
        c_ext_build_deps = ['Cython ~= 3.1.0a1']
```
It'd allow other versions below 3.1.1

lysnikolaou replied:
That would work for sure. There's still the problem [described here](https://github.com/aio-libs/propcache/pull/84#discussion_r1926906062) however.

rgommers replied:
`3.1.0a2-dev` in Cython master already has a number of thread safety fixes, mostyl minor but including one important one for fused types. Cython 3.1.x is in decent shape now, but `~=` or `>=` seems more useful here.

asvetlov reviewed (changes requested):
I would suggest waiting for non-alpha Cython release. A Release Candidate is ok, but alpha is not.

Dreamsorcerer commented:
> I would suggest waiting for non-alpha Cython release. A Release Candidate is ok, but alpha is not.

It's a compile-time experimental feature on cpython, I don't think anybody is expecting us to have a rock solid supported implementation. It's essentially a preview release at this point, so alpha is fine at this stage (and the only thing we can offer).

Dreamsorcerer commented:
The discussion above is to use it only on the 3.13t release. Stable Cython is used on other versions, so this has already been addressed.

asvetlov commented:
@Dreamsorcerer I'm sorry; I've missed that the new config is applied to 3.13t only.
There are no objections from me anymore.

lysnikolaou commented:
The problem with this is that there's sadly no environment marker for free-threading in the constraint file (though there are plans for it). This creates a conflict between the specified build requirement and the Cython constraints for regular 3.13. Any ideas on how you'd like to proceed here?

rgommers commented:
> This creates a conflict between the specified build requirement and the Cython constraints for regular 3.13. Any ideas on how you'd like to proceed here?

What all other projects that have released `cp313t` wheels do here is to set the requirements for 3.13 in `pyproject.toml` and other metadata file for the default CPython build (i.e., 3.0.11 now, not 3.1.0a1) and disable build isolation to produce `cp313t` wheels, installing Cython alpha/nightly in the wheel build job. I think that's the only good option at the moment, or something equivalent like pointing at a different constraints file (`requirements/cython_freethreading.txt` ?).

lysnikolaou commented:
Thanks @rgommers for the pointer! This is what an older verion of this PR did, though my understanding was that the here maintainers did not like that. If everyone is okay with this, I can certainly revisit that approach.

webknjaz commented on packaging/pep517_backend/_backend.py:
```diff
@@ -371,10 +371,7 @@ def get_requires_for_build_wheel(
-        'Cython ~= 3.0.0; python_version >= "3.12"',
-        'Cython; python_version < "3.12"',
-    ]
+    c_ext_build_deps = [] if is_pure_python_build else ['Cython ~= 3.0.11']
```
https://github.com/aio-libs/yarl/pull/1456/files#r1958383147

webknjaz commented:
> a different constraints file (`requirements/cython_freethreading.txt` ?).

@lysnikolaou I do like this approach. It'd be nice to granularly override it just for `3.13t`.

lysnikolaou commented:
I've update this PR to use the exact same approach as `yarl` (around building wheels). If someone could approve the workflows, that'd be really helpful.

bdraco commented:
Once done, feel free to reach out on matrix or discord any time you need the CI kicked

bdraco commented on packaging/pep517_backend/_backend.py:
```diff
@@ -371,10 +372,12 @@ def get_requires_for_build_wheel(
+    elif sysconfig.get_config_var('Py_GIL_DISABLED'):
+        c_ext_build_deps = ['Cython ~= 3.1.0a1']
+    else:
+        c_ext_build_deps = ['Cython ~= 3.0.11']
```
```suggestion
        c_ext_build_deps = ['Cython ~= 3.0.12']
```

bdraco commented on requirements/codspeed.txt:
```diff
@@ -0,0 +1,2 @@
+-r test.txt
+pytest-codspeed==3.1.2
```
```suggestion
pytest-codspeed==3.2.0
```

bdraco commented:
Thanks @lysnikolaou

I'm doing release now.  🤞 that we don't hit qemu issues again

bdraco commented:
Manually checked all the workflows. Build looks good. Going to release it out now
